vivarium/framework/components.py

```diff
@@ -1,63 +1,296 @@
+    """
+
+    component, *args = ast.iter_child_nodes(list(ast.iter_child_nodes(
+        list(ast.iter_child_nodes(ast.parse(desc)))[0]))[0])
+    component = _component_ast_to_path(component)
+    new_args = []
+    for arg in args:
+        parsed = False
+        try:
+            new_args.append(ast.literal_eval(arg))
+            parsed = True
+        except ValueError:
+            if isinstance(arg, ast.Call):
+                constructor, *constructor_args = ast.iter_child_nodes(arg)
+                constructor = constructors.get(constructor.id)
+                # NOTE: This currently precludes arguments other than strings.
+                # May want to release that constraint later.
+                if constructor and len(constructor_args) == 1 and isinstance(constructor_args[0], ast.Str):
+                    new_args.append(constructor(constructor_args[0].s))
+                    parsed = True
+                else:
+                    raise ParsingError('Invalid syntax: {}'.format(desc))
+        if not parsed:
+            raise ParsingError('Invalid syntax: {}'.format(desc))
+    return component, new_args
+
+
+def _prep_components(component_list: Sequence, constructors: Mapping[str, Callable]) -> Sequence:
+    """Transform component description strings into tuples of component callables and arguments the component may need.
+
+    Parameters
+    ----------
+    component_list
+        The component descriptions to transform
+    constructors
+        Dictionary of callables for creating argument objects
+
+    Returns
+    -------
+    List of component/argument tuples.
+    """
 
-def load(component_list):
     components = []
     for component in component_list:
         if isinstance(component, str):
             if '(' in component:
-                i = component.index('(')
-                args = literal_eval(component[i:])
-                if not isinstance(args, tuple):
-                    args = (args,)
-                component = component[:i]
+                component, args = _parse_component(component, constructors)
                 call = True
             else:
                 call = False
 
-            module_path, _, component_name = component.rpartition('.')
-            component = getattr(import_module(module_path), component_name)
+            component = _import_by_path(component)
+
+            for attr, val in inspect.getmembers(component, lambda a: not inspect.isroutine(a)):
+                constructor = constructors.get(val.__class__)
+                if constructor:
+                    setattr(component, attr, constructor(val))
 
             # Establish the initial configuration
             if hasattr(component, 'configuration_defaults'):
-                config.read_dict(component.configuration_defaults, layer='component_configs', source=module_path)
+                config.read_dict(component.configuration_defaults, layer='component_configs', source=component)
 
             if call:
```

Comment:
I also think it's worth revisiting the all components should be classes discussion.  I think that makes a lot of the code in this module much simpler.